### PR TITLE
upstream: force a full rebuild on host weight changes

### PR DESCRIPTION
**Pull request by mattklein123**
This will allow us to build load balancers that pre-compute data
structures based on host weights (for example using weighted queues),
to work around some of the deficiencies of EDF scheduling.

This behavior can be temporarily disabled by setting the
envoy.reloadable_features.upstream_host_weight_change_causes_rebuild
feature flag to false.

Fixes https://github.com/envoyproxy/envoy/issues/14360

Risk Level: Medium
Testing: New unit tests
Docs Changes: N/A
Release Notes: Added
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @tonya11en @snowp @antoniovicente

**Review by antoniovicente — APPROVED** [review]
Very nice.  Thanks for the fix!

**Comment by mattklein123** [comment]
(I think I'm going to hold this until after the release due to regression risk)

**Review by tonya11en — DISMISSED** [review]
LGTM as well 💯 . It's satisfying to see the results of all those discussions start to get merged!

File: test/common/upstream/eds_test.cc
Comment (tonya11en) [inline test/common/upstream/eds_test.cc:330]: nit - `endpoint` is misspelled.

**Comment by mattklein123** [comment]
@antoniovicente @tonya11en merged main and fixed typo, PTAL

**Review by antoniovicente — APPROVED** [review]
Looks good after merges.

**Pull request merged**
